sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -198,35 +196,70 @@ buildExport m =
 
 
+-- note: [Unmasking Asyncs]
+--
+-- It is possible to create unkillable asyncs. Behold:
+--
+-- ```
+-- a <- uninterruptibleMask_ $ do
+--     async $ do
+--         forever $ do
+--             threadDelay 10_000
+--             putStrLn "still alive"
+-- cancel a
+-- ```
+--
+-- The prior code block will never successfully cancel `a` and will be
+-- blocked forever. The reason is that `cancel` sends an async exception to
+-- the thread performing the action, but the `uninterruptibleMask` state is
+-- inherited by the forked thread. This means that *no async exceptions*
+-- can reach it, and `cancel` will therefore run forever.
+--
+-- This also affects `timeout`, which uses an async exception to kill the
+-- running job. If the action is done in an uninterruptible masked state,
+-- then the timeout will not successfully kill the running action.
+
 {- |
  The batch processor accepts spans and places them into batches. Batching helps better compress the data and reduce the number of outgoing connections
  required to transmit the data. This processor supports both size and time based batching.
 -}
 batchProcessor :: MonadIO m => BatchTimeoutConfig -> Exporter ImmutableSpan -> m Processor
 batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
   batch <- newIORef $ boundedMap maxQueueSize
-  workSignal <- newEmptyMVar
-  worker <- async $ loop $ do
-    req <-
-      liftIO $
-        timeout (millisToMicros scheduledDelayMillis) $
-          takeMVar workSignal
-    batchToProcess <- liftIO $ atomicModifyIORef' batch buildExport
-    res <- liftIO $ Exporter.exporterExport exporter batchToProcess
-
-    -- if we were asked to shutdown, quit cleanly after this batch
-    -- FIXME: this could lose batches if there's more than one in queue?
-    case req of
-      Just Shutdown -> throwE res
-      _ -> pure ()
+  workSignal <- newEmptyTMVarIO
+  shutdownSignal <- newEmptyTMVarIO
+  let workerAction = do
+        delay <- registerDelay (millisToMicros scheduledDelayMillis)
+        req <- atomically $ do
+          msum
+            [ const Flush <$> do
+                continue <- readTVar delay
+                check continue
+            , const Flush <$> takeTMVar workSignal
+            , const Shutdown <$> takeTMVar shutdownSignal
+            ]
+        batchToProcess <- atomicModifyIORef' batch buildExport
+        -- we mask async exceptions in this, so that a buggy exporter that
+        -- catches async exceptions won't swallow them. since we use
+        -- an interruptible mask, blocking calls can still be killed, like
+        -- `threadDelay` or `putMVar` or most file I/O operations.
+        --
+        -- if we've received a shutdown, then we should be expecting
+        -- a `cancel` anytime now.
+        res <- mask_ $ Exporter.exporterExport exporter batchToProcess
+
+        -- if we were asked to shutdown, quit cleanly after this batch
+        -- FIXME: this could lose batches if there's more than one in queue?
```

Comment:
I'd really love to see this addressed before merging.

Reply:
Hmm. I think we need to be able to distinguish between normal program exit and exceptional exit.

If we reinterpret `Shutdown` as "flush the queue until empty," instead of "shut down after the next flush," then we can have the following behavior:

On normal program exit, we want to block until the queue is flushed and everything is reported. Then this cleanup is done and we can go on to the next cleanup. Otherwise we lose metrics. We're not in a hurry, since we're exiting "normally."

But on exceptional exit, we *don't* want to block - we want to GTFO ASAP, so that we can run any other finalizers. In this case, we'd probably want to send a `Shutdown` to the worker thread, and exit immediately instead of `wait` or `cancel` or in any other way killing the thread. We should expect that the RTS will be reapin' us soon anyway, so we don't need to worry about anything too Weird happening.

If we cannot distinguish exceptional vs unexceptional, then we must accept the same behavior for both. In the "happy" case, the "exception-safe" behavior will allow the program to terminate as soon as the last finalizer is complete - potentially losing traces. In the "exception" path, the "happy" behavior will potentially block other resources from being cleaned up.

But allowing for exception vs unexceptional shutdown would require re-working the API a good bit more, I think.

Reply:
OK, I'm convinced it's probably not necessary for this PR.

That said, I am fairly convinced that we still want to fully process the queue on exceptional exit– if we lose sampled traces that show *why* a program is exiting exceptionally, it reduces the utility of the library for short-lived processes quite a bit . I would argue that since consumers of the library can specify max queue size, batch export size, and export timeout, they get to control the maximum amount of shutdown time required.

Reply:
Here's what the specification says about shutdown, FWIW: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shutdown-1

Reply:
> That said, I am fairly convinced that we still want to fully process the queue on exceptional exit– if we lose sampled traces that show why a program is exiting exceptionally, it reduces the utility of the library for short-lived processes quite a bit 

That makes sense to me. I suppose the *most* common case is that you start this up right near the very top of `main`, so there wouldn't be any other cleanup actions to run.